feat($compile): add more lifecycle hooks to directive controllers

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

Feature

**What is the current behavior? (You can also link to an open issue here)**

Missing lifecycle hooks for directive controllers

**What is the new behavior (if this is a feature change)?**

Add these lifecycle hooks

**Does this PR introduce a breaking change?**

Nope

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

This change adds in the following new lifecycle hooks, which map in some
way to those in Angular 2:
- `$onChanges(changesObj)` - Called whenever one-way bindings are updated. The `changesObj` is a hash whose keys
  are the names of the bound properties that have changed, and the values are an object of the form
  `{ currentValue: ..., previousValue: ... }`. Use this hook to trigger updates within a component such as
  cloning the bound value to prevent accidental mutation of the outer value.
- `$onDestroy` - Called on a controller when its containing scope is destroyed. Use this hook for releasing
  external resources, watches and event handlers.
- `$afterViewInit` - Called after this controller's element and its children been linked. Similar to the post-link
  function this hook can be used to set up DOM event handlers and do direct DOM manipulation.
  Note that child elements that contain `templateUrl` directives will not have been compiled and linked since
  they are waiting for their template to load asynchronously and their own compilation and linking has been
  suspended until that occurs.

Closes #14127
Closes #14030
Closes #14020
Closes #13991

mgol commented on docs/content/guide/component.ngdoc:
```diff
@@ -147,6 +147,27 @@ components should follow a few simple conventions:
+    function this hook can be used to set up DOM event handlers and do direct DOM manipulation.
+    Note that child elements that contain `templateUrl` directives will not have been compiled and linked since
+    they are waiting for their template to load asynchronously and their own compilation and linking has been
+    suspended until that occurs.
```
Is it also how it happens now if you use regular postLink functions in `module.directive`-defined directives?

petebacondarwin replied:
Yes

gkalpak replied:
TBH, I feel that this last limitation makes this hook pretty usel...ehm...much less useful than the corresponding ng2 hook :(

--
I assume, with the current implementation, this hook is very similar to the post-link function, which means that:

(a) It doesn't offer new functionality for ng1 users.
(b) It doesn't help much with migrating but for very simple components with inline templates. Even caching the templates won't help, right ?

These are just assumptions atm (I haven't looked at the code yet), but if they are true, I'm a little skeptical about this hook.

mgol replied:
In my current project I avoid these problems by not using `templateUrl` but relying on Webpack being able to import HTML files using the ES6 import syntax. Unfortunately you can't do things like that without any preprocessing done so it can't be a general solution.

petebacondarwin replied:
You are right @gkalpak, which is why I was suggesting that we call it `$afterLink` instead.
It does, though, have value for components, where there is no way right now to hook into the post-link function.

petebacondarwin replied:
I am becoming very keen on the webpack `template: require('some.template.html)` approach :-)

gkalpak replied:
Yeah, I only wish TypeScript had something similar :(
But in the meantime, I believe most developers follow the "best practice" of caching their templates with `$templateCache` in a build step. Besides a component doesn't know what it's children (or its grandchildren) do wrt template vs templateUrl.

I am afraid this will confuse people and create some headaches while migrating.

--
If we can't/won't make it work similarly to ng2, it might be btter to name is `$postLink` and just provide it as a way to access the post-link function while using the `.component()` helper, as @petebacondarwin suggested.

mgol commented on docs/content/guide/component.ngdoc:
```diff
@@ -147,6 +147,27 @@ components should follow a few simple conventions:
+  * `$onInit()` - Called on each controller after all the controllers on an element have been constructed and
+    had their bindings initialized (and before the pre &amp; post linking functions for the directives on
+    this element). This is a good place to put initialization code for your controller.
+  * `$onChanges(changesObj)` - Called whenever one-way bindings are updated. The `changesObj` is a hash whose keys
```
I love it (as well as `$onDestroy`); more reasons to stop injecting `$scope` into controllers.

mgol commented on src/ng/compile.js:
```diff
@@ -3090,6 +3119,27 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
 
+      function recordChanges(key, currentValue, previousValue) {
+        if (isFunction(destination.$onChanges)) {
+          // If we have not already schedule the onChanges hook then do so now
```
`s/schedule/scheduled/`

mgol commented:
I'd add tests that ensure `$onChanges` is **not** called on changes in other kinds of bindings.

petebacondarwin commented:
Fixed the comment typo and added tests @mgol PTAL

gkalpak commented on docs/content/guide/component.ngdoc:
```diff
@@ -147,6 +147,27 @@ components should follow a few simple conventions:
+    cloning the bound value to prevent accidental mutation of the outer value.
+  * `$onDestroy` - Called on a controller when its containing scope is destroyed. Use this hook for releasing
+    external resources, watches and event handlers.
+  * `$afterViewInit` - Called after this controller's element and its children been linked. Similar to the post-link
```
_been linked_  -->  _have been linked_

mgol commented:
LGTM (once the typo @gkalpak noticed is fixed)

gkalpak commented on src/ng/compile.js:
```diff
@@ -293,9 +293,21 @@
+ * * `$onInit()` - Called on each controller after all the controllers on an element have been constructed and
  *   had their bindings initialized (and before the pre &amp; post linking functions for the directives on
  *   this element). This is a good place to put initialization code for your controller.
+ * * `$onChanges(changesObj)` - Called whenever one-way bindings are updated. The `changesObj` is a hash whose keys
```
Nitpick: I would put the `<` symbol there somewhere, because `@` bindings are also called "one-way" in many blogsposts/tutorials and this might confuse users.

petebacondarwin replied:
OK

gkalpak commented on src/ng/compile.js:
```diff
@@ -293,9 +293,21 @@
+ *   cloning the bound value to prevent accidental mutation of the outer value.
+ * * `$onDestroy` - Called on a controller when its containing scope is destroyed. Use this hook for releasing
+ *   external resources, watches and event handlers.
+ * * `$afterViewInit` - Called after this controller's element and its children been linked. Similar to the post-link
```
Again: _**have** been linked_

gkalpak commented on src/ng/compile.js:
```diff
@@ -3101,6 +3130,27 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+
+      function triggerOnChangesHook() {
+        // We must run this hook in an apply since the $$postDigest runs outside apply
+        scope.$apply(function() {
```
Doesn't this mean there will be one extra digest for each controller (that has at least one changed binding) per digest ?

petebacondarwin replied:
Yes you are right.
The alternative is to use `$applyAsync` but this has the downside of delaying the call to `$onChanges` by maybe 100ms.

What is better?

gkalpak replied:
I think we could move the login that calls `triggerOnChangeHook` inside `$compile` (I mean make it a compiler-local thing) and store the changes for **each** controller there, so we only trigger one extra `$digest`.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3515,6 +3515,202 @@ describe('$compile', function() {
+          var d2Controller = element.find('d2').controller('d2');
+          var d3Controller = element.find('d3').controller('d3');
+
+          expect([d1Controller.count, d2Controller.count, d3Controller.count]).toEqual([0,0,0]);
```
`[0,0,0]` ? What were you doing JSCS ? How did you let this happen ? :disappointed:

petebacondarwin replied:
That style of array appears throughout this file

gkalpak replied:
That's why I'm so disappointed with JSCS :stuck_out_tongue:

gkalpak commented:
I am also wondering if applying `$onChange` to `@` bindings would be closer to ng2 (iirc it will).

gkalpak commented on docs/content/guide/component.ngdoc:
````diff
@@ -147,6 +147,30 @@ components should follow a few simple conventions:
       }
     ```
 
+- **Components have a well-defined lifecycle
````
Missing closing `**`.

gkalpak commented on docs/content/guide/component.ngdoc:
```diff
@@ -147,6 +147,30 @@ components should follow a few simple conventions:
+    cloning the bound value to prevent accidental mutation of the outer value.
+  * `$onDestroy` - Called on a controller when its containing scope is destroyed. Use this hook for releasing
+    external resources, watches and event handlers.
+  * `$postLink` - Called after this controller's element and its children been linked. Similar to the post-link
```
_been_  -->  _have been_

gkalpak commented on src/ng/compile.js:
```diff
@@ -293,9 +293,21 @@
+ *   whose keys are the names of the bound properties that have changed, and the values are an object of the form
+ *   `{ currentValue: ..., previousValue: ... }`. Use this hook to trigger updates within a component such as
+ *   cloning the bound value to prevent accidental mutation of the outer value.
+ * * `$onDestroy` - Called on a controller when its containing scope is destroyed. Use this hook for releasing
```
There is no `()`  after `$onDestroy` (and `$postLink`) for consistency.

gkalpak replied:
It might be a good idea to document (here or elsewhere) that parent->child execution order.

gkalpak commented on src/ng/compile.js:
```diff
@@ -474,7 +486,7 @@
  *   * `iElement` - instance element - The element where the directive is to be used. It is safe to
  *     manipulate the children of the element only in `postLink` function since the children have
- *     already been linked.
+ *     already have been linked.
```
Wrong _been_. We don't need _have_ here.

petebacondarwin replied:
Doh!

gkalpak commented on src/ng/compile.js:
```diff
@@ -1215,6 +1227,24 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+    // This function is called in a $$postDigest to trigger all the onChanges hooks in a single digest
+    function flushOnChangesQueue() {
+      // We must run this hook in an apply since the $$postDigest runs outside apply
+      $rootScope.$apply(function() {
```
I would only run a digest if there are actually entries in `onChangesQueue`.

petebacondarwin replied:
The `flushOnChangesQueue()` is only called if there are entries

gkalpak replied:
Right :grin:

gkalpak commented on src/ng/compile.js:
```diff
@@ -1215,6 +1227,24 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+    function flushOnChangesQueue() {
+      // We must run this hook in an apply since the $$postDigest runs outside apply
+      $rootScope.$apply(function() {
+        for (var i = 0, ii = onChangesQueue.length; i < ii; ++i) {
```
What about changes detect during this digest ?

petebacondarwin replied:
Good call...
All the items in the onChanges queue will be flushed before the new digest is triggered.

The sequence looks like:

* some change 1
* some change 2
* $digest
  * schedule flushOnChangesQueue
  * queue onChanges 1
  * queue onChanges 2
* $$postDigest
  * flushOnChangesQueue
    * $apply
      * trigger onChanges 1
        * some change 1a
      * trigger onChanges 2
        * some change 2a
      * $digest
        * schedule onChangesFlush
        * queue onChanges 1
        * queue onChanges 2
      * $$postDigest
        * flushOnChangesQueue
          * $apply
            * trigger onChanges 1
            * trigger onChanges 2
          * $digest

I think this means that we are OK (except for the potential for deep recursion). I'll make a test

petebacondarwin replied:
Added a test @gkalpak PTAL

gkalpak replied:
So, this looks good apart from the infinite loop issue.

Ideally, there should be some mechanism in place to prevent freezing the whole app, because someone misused `$onChanges` (a TTL-like mechanism maybe).
If not (e.g. it turns out to be too complicated), we should at least document very clearly the dangers of making changes that would trigger another `$onChanges` round. from inside an `$onChanges()` callback.

gkalpak replied:
So, this looks good apart from the infinite loop issue.

Ideally, there should be some mechanism in place to prevent freezing the whole app, because someone misused `$onChanges` (a TTL-like mechanism maybe).
If not (e.g. it turns out to be too complicated), we should at least document very clearly the dangers of making changes that would trigger another `$onChanges` round. from inside an `$onChanges()` callback.

petebacondarwin replied:
It is not an infinite loop, it is an infinite recursion, which will trigger a `RangeError` exception that will prevent the browser from freezing and indicate to the developer that they have done a bad thing.

Added a test to demonstrate.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3515,6 +3515,284 @@ describe('$compile', function() {
+
+          // Update val to trigger the onChanges
+          $rootScope.$apply('a = 42');
+          // Now we the change should have the real previous value (undefined), not the intemediate one (42)
```
-we
intemediate  -->  intermediate

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3515,6 +3515,284 @@ describe('$compile', function() {
+
+          // Update val to trigger the onChanges
+          $rootScope.$apply('a = 7');
+          // Now we the change should have the real previous value (126), not the intemediate one, (91)
```
Same typos as above.

gkalpak commented on src/ng/compile.js:
```diff
@@ -293,9 +293,23 @@
+ * * `$onInit()` - Called on each controller after all the controllers on an element have been constructed and
  *   had their bindings initialized (and before the pre &amp; post linking functions for the directives on
  *   this element). This is a good place to put initialization code for your controller.
+ * * `$onChanges(changesObj)` - Called whenever one-way (`<`) bindings are updated. The `changesObj` is a hash
```
Now we need to change this again to also mention `@` :grin:

dcherman commented on src/ng/compile.js:
```diff
@@ -3101,6 +3152,33 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+        if (isFunction(destination.$onChanges) && currentValue !== previousValue) {
+          // If we have not already scheduled the top level onChangesQueue handler then do so now
+          if (!onChangesQueue) {
+            scope.$$postDigest(flushOnChangesQueue);
```
Why `$$postDigest` versus `$evalAsync`?  Didn't pull this down to test, but `$evalAsync` would feel like the more correct method to use here so that we continue to live within the same digest cycle and don't accidentally violate the max TTL due to handling record changes

petebacondarwin replied:
We want all the changes in the current digest to settle before we trigger the onChanges hook.
For this to happen we need the current digest to have completed, which is why postDigest is a good place. Perhaps you can convince me otherwise?

dcherman replied:
No need anymore FWIW.  The addition of the other TTL mechanism mitigated most of my concern, though it'd be sweet if the existing TTL could have been passed along rather then implementing another one so that the max TTL can be defined in one place.

petebacondarwin replied:
Hi @doberman can you clarify what you mean here?
Are you saying that the use of EvalAsync is not needed any more?

I did consider piggybacking on the rootscope TTL but since it is not actually related to the digest cycle I  thought we should keep them independent

dcherman commented on src/ng/compile.js:
```diff
@@ -2360,10 +2392,16 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+          }
+          if (isFunction(controllerInstance.$onDestroy)) {
+            controllerScope.$on('$destroy', function callOnDestroyHook() {
+              controllerInstance.$onDestroy();
```
This might be a ridiculous edge case, but is there any concern about removal of the hook during the lifetime of the controller?  `$onDestroy` may not necessarily exist on the instance in the future even though it does right now.

petebacondarwin replied:
For the moment I don't want to support dynamic changes to the component hook methods: neither lazy definition nor removal. If this is a really common use case we could consider implementing it later without introducing a breaking change.

gkalpak commented:
LGTM with https://github.com/angular/angular.js/pull/14302/files#r57215106 (somehow) addressed
:+1:

petebacondarwin commented:
I've added time-to-live constraints for onChanges hooks, which should address @gkalpak's last concern

Narretz commented on docs/content/error/$compile/infchng.ngdoc:
````diff
@@ -0,0 +1,24 @@
+<c1 prop="a" on-change="a = -a"></c1>
+```
+
+```js
````
Can you add the component definition object? Makes it easier to understand if you know the bindings are `bindings: {'prop': '<', onChange: '&'}`

petebacondarwin replied:
Done

gkalpak commented on src/ng/compile.js:
```diff
@@ -1221,6 +1221,34 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * Increasing the TTL could have performance implications, so you should not change it without proper justification.
+   *
+   * @param {number} limit The number of `$onChanges` hook iterations.
+   */
```
Missing `@returns` docs.

petebacondarwin replied:
Done

gkalpak commented on src/ng/compile.js:
```diff
@@ -1221,6 +1221,34 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+  this.onChangesTtl = function(value) {
+    if (arguments.length) {
+      TTL = value;
+    }
```
I suggest returning `this` for consistency. (And then also document it :smiley:)

petebacondarwin replied:
Done

gkalpak commented on src/ng/compile.js:
Shouldn't it somehow be reset (at some point) ?

gkalpak commented on src/ng/compile.js:
```diff
@@ -1207,6 +1221,36 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * Increasing the TTL could have performance implications, so you should not change it without proper justification.
+   *
+   * @param {number} limit The number of `$onChanges` hook iterations.
+   * @returns {number|object} the current limit (or `this` if called as a setting for chaining)
```
_setting_  -->  _setter_

gkalpak commented on src/ng/compile.js:
```diff
@@ -1215,6 +1259,31 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+
+
+
+    var onChangesTtl;
```
With the current implementation couldn't we initialize `onChangeTtl` here ?

petebacondarwin replied:
true

gkalpak commented on src/ng/compile.js:
```diff
@@ -1215,6 +1259,31 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+    // This function is called in a $$postDigest to trigger all the onChanges hooks in a single digest
+    function flushOnChangesQueue() {
+      if (!(onChangesTtl--)) {
+        throw $compileMinErr('infchng', '{0} $onChanges() iterations reached. Aborting!\n', TTL);
```
I believe we should increment `onChangesTtl` before throwing.

petebacondarwin replied:
OK!

petebacondarwin replied:
Actually we need to reset it to the original value TTL

gkalpak commented on src/ng/compile.js:
```diff
@@ -1215,6 +1259,32 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+    // This function is called in a $$postDigest to trigger all the onChanges hooks in a single digest
+    function flushOnChangesQueue() {
+      if (!(onChangesTtl--)) {
+        onChangesTtl = TTL;
```
I don't think this is correct. We should just `onChangesTtl++`, because the exception will be caught in this digest's `catch` block, so the rest of the ++ (on https://github.com/angular/angular.js/pull/14302/files#diff-a732922b631efed1b9f33a24082ae0dbR1284) will be executed anyway.

gkalpak commented on src/ng/compile.js:
This comment is no longer needed.

petebacondarwin committed:
feat($compile): add more lifecycle hooks to directive controllers

This change adds in the following new lifecycle hooks, which map in some
way to those in Angular 2:

 * `$onChanges(changesObj)` - Called whenever one-way bindings are updated. The `changesObj` is a hash whose keys
   are the names of the bound properties that have changed, and the values are an object of the form
   `{ currentValue: ..., previousValue: ... }`. Use this hook to trigger updates within a component such as
   cloning the bound value to prevent accidental mutation of the outer value.
 * `$onDestroy` - Called on a controller when its containing scope is destroyed. Use this hook for releasing
   external resources, watches and event handlers.
 * `$postLink` - Called after this controller's element and its children been linked. Similar to the post-link
   function this hook can be used to set up DOM event handlers and do direct DOM manipulation.
   Note that child elements that contain `templateUrl` directives will not have been compiled and linked since
   they are waiting for their template to load asynchronously and their own compilation and linking has been
   suspended until that occurs.

Closes #14127
Closes #14030
Closes #14020
Closes #13991
Closes #14302
